radicle-httpd/src/raw.rs

```diff
@@ -0,0 +1,54 @@
+use std::sync::Arc;
+
+use axum::extract::State;
+use axum::response::IntoResponse;
+use axum::routing::get;
+use axum::Router;
+
+use radicle::prelude::Id;
+use radicle::profile::Profile;
+use radicle::storage::git::paths;
+use radicle_surf::{Oid, Repository};
+
+use crate::api::axum_extra::Path;
+use crate::api::error::Error;
+
+pub fn router(profile: Arc<Profile>) -> Router {
+    Router::new()
+        .route("/:project/:sha/*path", get(file_handler))
+        .with_state(profile)
+}
+
+async fn file_handler(
+    Path((project, sha, path)): Path<(Id, Oid, String)>,
+    State(profile): State<Arc<Profile>>,
+) -> impl IntoResponse {
+    let storage = &profile.storage;
+    let repo = Repository::open(paths::repository(storage, &project))?;
+    let blob = repo.blob(sha, &path)?;
+
+    Ok::<_, Error>(blob.content().to_owned())
```

Comment:
If there was a method like `blob.into() -> Vec<u8>` we could return it here without doing a copy.

Reply:
> If there was a method like `blob.into() -> Vec<u8>` we could return it here without doing a copy.

I don't understand as `to_owned()` is basically doing this, i.e. returning `Blob<Vec<u8>>`.  I wanted to clarify: 

- `repo.blob` returns a `Blob<BlobRef<'a>>` that does not copy or allocate anything. It reuses the `git2::Blob<'a>` internally. 

- `content()` return `&[u8]` , so does not allocate or copy.

- `to_owned()` returns `Blob<Vec<u8>>` , which makes a copy.

Reply:
issue is not copying here, issue is serving very big files. if we serve a 200MB binary, that'll use up 200MB of memory on server, 5 people connecting at the same time means a small box will run OOM. streaming won't have this issue.

Reply:
Understood, this is my concern too, and it was the reason we refactored `repo.blob()` to return a reference to underlying git2::blob by default. 

However, if `git2::blob` already consumes big memory, we will still have a problem.  Like I mentioned in another comment, another approach is to save large blobs to a temp file on disk, then do streaming from the file.  After its done, we will delete the temp file.

Reply:
Ah my bad, I thought `Blob` internally kept an owned vector, that's why I suggested that.

How about we check the blob length; if it's smaller than `1024` bytes, we return a copy via `to_owned`. If it's larger, we return an error code [413](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/413)

Reply:
I'd prefer if we add an API to stream from disk for blobs > `4096`

Reply:
`4096` (i.e. 4KB) bytes seems too low to me. How about 4MB (or 400KB) as the threshold?  For the error code, it can work as a stopgap, but we probably need to support streaming soon.

Reply:
I opened a PR for the basic support of blob streaming using a tmp file: https://github.com/radicle-dev/heartwood/pull/347 .

Reply:
> 
> How about we check the blob length; if it's smaller than `4096` bytes, we return a copy via `to_owned`. If it's larger, we return an error code [413](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/413)

I think we don't gain much in current implementation, because the main concern is memory usage on server side, if we already have the blob in `&[u8]`, then memory has been allocated. Bandwidth is cheap, and not much of a concern. (Unless being DDoS'd)

We can revisit this issue after #347.